core/src/main/java/io/questdb/griffin/engine/functions/bool/InSymbolCursorFunctionFactory.java

```diff
@@ -168,41 +164,45 @@ public void init(SymbolTableSource symbolTableSource, SqlExecutionContext execut
             this.valueSet = valueSet;
 
             RecordCursorFactory factory = cursorArg.getRecordCursorFactory();
-            cursor = factory.getCursor(executionContext);
+            try (RecordCursor cursor = factory.getCursor(executionContext)) {
+                final Record record = cursor.getRecord();
+                StringSink sink = Misc.getThreadLocalSink();
+                while (cursor.hasNext()) {
+                    CharSequence value = func.get(record, 0, sink);
+                    if (value == null) {
+                        this.valueSet.addNull();
+                    } else {
+                        int toIndex = this.valueSet.keyIndex(value);
+                        if (toIndex > -1) {
+                            this.valueSet.addAt(toIndex, Chars.toString(value));
+                        }
+                    }
+                }
+            }
         }
 
         @Override
-        public void initCursor() {
-            if (cursor != null) {
-                BinaryFunction.super.initCursor();
-                buildValueSet();
-                cursor = Misc.free(cursor);
-            }
+        public boolean isThreadSafe() {
+            return valueArg.isThreadSafe();
         }
 
         @Override
-        public boolean isThreadSafe() {
-            return false;
+        public void offerStateTo(Function that) {
+            if (that instanceof StrInCursorFunc) {
+                StrInCursorFunc thatF = (StrInCursorFunc) that;
+                thatF.valueSetB.clear();
+                thatF.valueSetB.addAll(valueSet);
+                thatF.valueSet = valueSetA;
+                thatF.stateInherited = this.stateShared = true;
+            }
+            BinaryFunction.super.offerStateTo(that);
```

Comment:
Looks like this call should be in else branch. The same question for other modified functions.

Reply:
i think this is correct, the intent is to propagate offer to the left argument if this argument ever needs state sharing.

Reply:
I see. This will also call `offerStateTo()` on `cursorArg`, but probably that's not a big concern as it'll be ignored now.

BTW shouldn't we make the casts unconditional? Otherwise, if `offerStateTo()` is ever called for different function trees, it may lead to "false positive" state copying.

Reply:
sorry, i did not get it, which casts?